### DISCUSSION TITLE
Using @collection.* to prevent creation of two rows with the same column value

### DISCUSSION THREAD

**Discussion by Xenograph** [pocketbase/pocketbase#5690] [Q&A] [answered]
My create API rule for the `users` collections is

```
invite != null && (invite.email = null || invite.email = email) && (@collection.users.id ?!= id && @collection.users.invite != invite)
```

This is meant to ensure that you cannot create a new account with an invite code that has already been used (note that invite is a foreign key to an invites table). I do not want to use a unique constraint here since eventually I would like certain conditions to support multi-use invites. 

With the rule as above, I cannot insert any rows. On the client I see `{"code":400,"message":"Failed to create record.","data":{}}` and on the server error log I see `DrySubmit create rule failure: sql: no rows in result set`.

Any ideas what's going on?

**Comment by ganigeorgiev** [comment]
The Create API rule at the moment executes AFTER a dry submit of the record (aka. the record will be in the database) and if the rule fails the transaction is rollbacked.

With that said, you are excluding the current id from the check so I guess it should work but the last condition needs also to be prefixed with `?` because otherwise it will apply a "match-all" type of constraint.

If that doesn't help, please export your collections configuration as json (_Admin UI > Settings > Export collections_).

**Comment by Xenograph** [comment]
I tried it with `?!=` on the last condition as well. Unfortunately, this doesn't produce intended behavior either. Instead of an error, it just seems to have no effect. ie. I can create as many rows as I want with the same invite.

Attached is the collection config for the `users` and `invites` collections. 
[pb_schema (1).json](https://github.com/user-attachments/files/17396253/pb_schema.1.json)

**Reply by ganigeorgiev** [reply, marked as answer]
On second read of your requirements and testing with the above schema (_I only removed the "house" field_), the proposed rule will not work because if I'm not mistaken what you want is a match-all constraint for the self-join but excluding the currently created user and this is not possible. When using the `?` prefix it allows users creation because most likely you have at least one user that satisfy the condition.

As mentioned earlier this issue is caused because of the "dry insert" of the created record which affects the self-join.

Since I don't exactly understand what you have in mind by _"I would like certain conditions to support multi-use invites"_ and a UNIQUE constraint is not an option then one possible hacky workaround could be to create a helper view collection that will contain for example how many time an invite code has been used and you could join that in your users collection and compare the used times (it should be + 1 because of the Create API rule dry insert).

An example "usedInvites" View collection query could be something like:

```sql
SELECT invites.id, COUNT(*) times
FROM invites
INNER JOIN users ON users.invite = invites.id
GROUP BY invites.id
```

And then your API rule could be:

```js
invite != "" &&
(invite.email = "" || invite.email = email) &&
// +1 because the Create API rule is checked after "dry insert"
(@collection.usedInvites.id ?= invite && @collection.usedInvites.times ?= 1)
```


_I have it in my todo to think of way to change and normalize the Create API rule behavior but that will be after the upcoming v0.23.0 as it ended up including too many changes and I don't want to block it further._

**Reply by Xenograph** [reply]
Thanks for the detailed answer. I hope that the filter rules will eventually become more flexible and support cases such as these.

**Reply by ganigeorgiev** [reply]
_Just in case someone stumble here: the dry submit behavior is discussed to change in v0.24+ per https://github.com/pocketbase/pocketbase/discussions/6073._
